### PR TITLE
[clang][Sema] Emit more specific diagnostic for auto in lambda before C++14 (#46059)

### PR DESCRIPTION
**Pull request by weltschildkroete**
Namely, we specify that `auto` in a lambda parameter is a C++14 extension in the error message, which now reads:

`'auto' not allowed in lambda parameter before C++14`

This does not change the behavior for `decltype(auto)` and `__auto_type` though.

### CODE REVIEW

**Comment by cor3ntin** [comment]
Thanks for this patch!

Our policy is to not mix functional changes and formatting changes.
Could you change the PR to have only the changes you made, wit no formatting change?
You can format your patch with `git clang-format` which only formats the lines that you changed.

**Comment by weltschildkroete** [comment]
> Thanks for this patch!

Thank you for the feedback and for your patience. :-)

It should be fixed now (hopefully!)

File: clang/include/clang/Basic/DiagnosticSemaKinds.td [hunk 1]
```diff
@@ -2393,7 +2393,7 @@ def err_auto_not_allowed : Error<
   "|in type allocated by 'new'|in K&R-style function parameter"
   "|in template parameter|in friend declaration|in function prototype that is "
   "not a function declaration|in requires expression parameter"
-  "|in array declaration}1">;
+  "|in array declaration|in lambda parameter until C++14}1">;
```
Comment (cor3ntin) [inline clang/include/clang/Basic/DiagnosticSemaKinds.td:2396]:
```suggestion
  "|in array declaration|in lambda parameter before C++14}1">;
```

To be consistent with other diagnostics

**Review by cor3ntin — APPROVED** [review]
A small nit, otherwise LGTM

I'm not sure it's strictly necessary here, but you could add a release note in clang/docs/ReleaseNotes.rst - under the diagnostics improvement section

File: clang/include/clang/Basic/DiagnosticSemaKinds.td (hunk 1 above)
Comment (weltschildkroete) [reply clang/include/clang/Basic/DiagnosticSemaKinds.td]: Thanks, done.

**Comment by weltschildkroete** [comment]
LMK if the release note sounds good. If so, could you land this patch for me? Please use "Leonardo Duarte weltschildkroete@gmail.com" to commit the change. Thanks!

File: clang/lib/Sema/SemaType.cpp [hunk 2]
```diff
@@ -3605,8 +3605,11 @@ static QualType GetDeclSpecTypeForDeclarator(TypeProcessingState &state,
         Info = &SemaRef.InventedParameterInfos.back();
       } else {
         // In C++14, generic lambdas allow 'auto' in their parameters.
-        if (!SemaRef.getLangOpts().CPlusPlus14 || !Auto ||
-            Auto->getKeyword() != AutoTypeKeyword::Auto) {
+        if (!SemaRef.getLangOpts().CPlusPlus14 && Auto &&
+            Auto->getKeyword() == AutoTypeKeyword::Auto) {
+          Error = 24;
```
Comment (shafik) [inline clang/lib/Sema/SemaType.cpp:3610]: A lot of the other cases have comments explaining what diagnostic the code pertains to. See 20 or 7 below for example. 

This is not really great but at least we have a guide to which diagnostic we intend to emit.

Comment (weltschildkroete) [reply clang/lib/Sema/SemaType.cpp]: Makes sense. I added some similar comments as well.

**Comment by cor3ntin** [comment]
@shafik I'll let you have the final say :)

**Review by shafik — APPROVED** [review]

**Pull request merged**